_pages/exercises/AutonomousCars/end_to_end_visual_control.md

```diff
@@ -0,0 +1,414 @@
+</p>
+
+### 4. Model Training and Evaluation Pipeline
+
+#### i) Data Preprocessing
+
+<p style="text-align:justify">
+Data preprocessing ensures that all inputs to the model are in a clean, consistent, and usable format. 
+The <strong>Combine Circuit Dataset</strong> is already balanced.
+</p>
+
+<p style="text-align:justify">
+However, the <strong>Simple Circuit Dataset</strong> is <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">imbalanced</code>. Training on such <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">imbalanced</code> data can cause the model to favor majority classes and perform poorly.
+To address this, you should first <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">categorize</code> the samples into their respective classes and then apply data balancing techniques, such as:
+
+  <ul>
+    <li><strong>Undersampling:</strong> Reducing the number of majority-class samples.</li>
+    <li><strong>Oversampling:</strong> Increasing the number of minority-class samples, sometimes by duplicating or augmenting them.</li>
+  </ul>
+</p>
+
+<p style="text-align:justify">
+In the preprocessing stage, images are first loaded from the dataset folders and paired with their corresponding labels from <strong>CSV files</strong> that contain control values such as <strong>linear</strong> and <strong>angular velocity</strong>. To ensure consistency, all images are <strong>resized</strong> to a fixed dimension and <strong>normalized</strong> by scaling pixel values to a 
+standard range, which improves training stability. The <strong>labels</strong> are extracted directly from the CSV files and linked to the appropriate images. To enhance <strong>robustness</strong> and <strong>generalization</strong>, <strong>data augmentation techniques</strong> 
+such as <strong>flipping</strong>, <strong>rotation</strong>, and <strong>brightness adjustment</strong> are applied, 
+creating diverse variations of the training data.
+</p>
+
+#### ii) Dataset Splitting
+
+<p style="text-align:justify">
+The <strong>Simple Circuit dataset</strong> is divided into three sets: <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">training</code> and <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">testing</code>. The training set contains the majority of the data and is used to teach the model to recognize patterns and learn the task. The testing set is the data used during training to fine-tune hyper-parameters, prevent over-fitting, and monitor the model’s performance on unseen data.
```

Comment:
This is a bit misleading. We have two sets (training and testing), not three. Ideally, we would have train, validation and test. In our case, we will test directly against the simulation in RoboticsAcademy, and so the "testing" split in your HuggingFace dataset serves actually as validation. Can you clarify that?